### PR TITLE
Clustering

### PR DESCRIPTION
**Pull request by afhuertass**
Clustering feature.

### CODE REVIEW

File: components/cluster.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        return []
+    #df['DateTime'] = df.apply(lambda x: get_datetime(x), axis = 1)
+    #df["TimeStamp"] = df.DateTime.values.astype(np.int64) / 10 ** 9
+
+    data = df[[ "TimeStamp" , "EVENTID" , "TimeStamp" , "MAGNITUDE" , "LATITUDE" , "LONGITUDE" ] ].values
+    edges = compute_edges_numba( data )
+    return edges
+
+
+@jit( nopython=True )
```
Comment (niklasmaki) [inline components/cluster.py]: What does this do?

File: components/cluster.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    items.append( html.H1("Clustering" ) )
+    items.append( html.H3("Select a date range"))
+    items.append( 
+
+    html.Div([
+    dcc.DatePickerRange(
+        id='date-pick',
+        min_date_allowed=mindate,
+        max_date_allowed=maxdate,
+    ),
```
Comment (niklasmaki) [inline components/cluster.py]: As we talked, this should probably be its own component.

File: components/cluster.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    G.add_edges_from( edges )
+
+    th = 1e3
+    to_remove = [ ]
+    for e in G.edges:
+        #print( e )
+        w = G.edges[ e ]["w"]
+        #print( 1.0/w )
+        if w  >= th :
+            print("remove")
```
Comment (niklasmaki) [inline components/cluster.py]: Maybe comment this one out as well, or is it needed?

**Review by niklasmaki — CHANGES_REQUESTED** [review]
So a few more comments about the changes in addition to the ones marked within the code:

- So apparently your linter was not working so there are a lot of PEP8 violations. I think the easiest way to fix this is open a file the linter complains about with VSCode and press ALT+SHIFT+F. VSCode should ask you if you want to install autopep8 formatter which should fix most of the issues. You can also take a look at the [PEP8 specification](https://www.python.org/dev/peps/pep-0008/).
- None of your functions have docstrings. You should add those at least to the most important ones, i.e. those that are called from other files. You can take a look at for example uploader.py for the docstring format.

After the changes I'm gonna review the code again.

**Comment by niklasmaki** [comment]
And also you should fix the merge conflict (again)

File: components/cluster.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    eq_data = earthquake_data.get_earthquake_data(session_id)
+    return eq_data
+
+
+def compute_edges(df):
+    """Return the list of edges to be used to build the clustering graph."""
+    if df.empty:
+        return []
+
+    columns = ["TimeStamp", "EVENTID", "TimeStamp", "MAGNITUDE", "LATITUDE", "LONGITUDE"]
```
Comment (niklasmaki) [inline components/cluster.py]: Why is there TimeStamp twice?

File: components/cluster.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    data = df[columns].values
+    edges = compute_edges_numba(data)
+    return edges
+
+
+@jit(nopython=True)
+def compute_edges_numba(data):
+    """Return the list of edges to be used to build the clustering graph.
+    The edges are computed used numba compiler for infinite power
+    Keyword arguments:
```
Comment (niklasmaki) [inline components/cluster.py:63]: There should be an empty line above "Keyword arguments"

File: components/cluster.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+from app import app
+import datetime
+from utils import earthquake_data
+import math
+from numba import njit, jit
+import plotly.graph_objects as go
+import networkx as nx
+from datetime import datetime as dt
+import dash_bootstrap_components as dbc
```
Comment (niklasmaki) [inline components/cluster.py]: Imports should be divided into 3 categories: 

1. Standard library imports.
2. Related third party imports.
3. Local application/library specific imports.

These categories should be separated by empty lines. See [PEP8](https://www.python.org/dev/peps/pep-0008/#imports) for more information.

File: utils/earthquake_data.py [hunk 7]
```diff
@@ -4,18 +4,50 @@
...
 
     def __init__(self, catalog_type, data):
         self.catalog_type = catalog_type
         self.data = data
+        self.compute_datetime()
+
+    def compute_datetime(self):
+        if not self.data.empty:
+            self.data["DateTime"] = self.data.apply(lambda x: get_datetime(x), axis=1)
+            self.data["TimeStamp"] = self.data.DateTime.values.astype(np.int64)/10**9
```
Comment (niklasmaki) [inline utils/earthquake_data.py]: PEP8 lint complains that these lines are too long. Could they be divided into shorter lines? autopep8 does this automatically if you managed to install it.

File: views/clusterview.py [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+from app import app
+from components import uploader
+from components import cluster
+
+
+def get_layout( session_id ):
+
+    return cluster.get_component( session_id )
+
```
Comment (niklasmaki) [inline views/clusterview.py]: Multiple PEP8 problems here as well.

**Review by niklasmaki — CHANGES_REQUESTED** [review]
Still some more small suggestions :)

File: components/cluster.py (hunk 2 above)
Comment (niklasmaki) [reply components/cluster.py]: This will be done in a separate task.

File: components/cluster.py (hunk 1 above)
Comment (niklasmaki) [reply components/cluster.py]: So as I understand it, this jits the method to lower-level code that's more efficient (has to do with numba).

File: utils/earthquake_data.py (hunk 7 above)
Comment (afhuertass) [reply utils/earthquake_data.py]: Should be fixed in the newest commit. The second line I couldn't get any shorter.

File: views/clusterview.py (hunk 8 above)
Comment (afhuertass) [reply views/clusterview.py]: You really do hate spaces. Fixed.

**Review by niklasmaki — APPROVED** [review]
I think all of the linting problems are solved now. And other changes look good as well, so feel free to merge 🙂

**Comment by niklasmaki** [comment]
Although the conflicts must be fixed before merging...

**Comment by afhuertass** [comment]
how can I solve the conflicts my local machine? so I can run the linter and make sure that everything stays on place.

**Comment by niklasmaki** [comment]
In your local environment:

1) `git checkout clustering`
2) `git merge master`
3) Resolve the conflicts and then commit as usual. Then just `git push` and your fixes are shown here.

**Comment by afhuertass** [comment]
I did a couple of large changes to merge the branches, I believe it is worth you @niklasmaki  take a look and make sure everything is fine before the merge :ok_hand:

File: components/cluster.py [hunk 9]
```diff
@@ -23,15 +23,43 @@ def get_data(session_id):
     return eq_data
 
 
-def compute_edges(df):
-    """Return the list of edges to be used to build the clustering graph."""
+def compute_edges(data, df):
```
Comment (niklasmaki) [inline components/cluster.py]: It's kind of weird that the df is given as an argument when the only thing it's used for is to check whether it's empty. Is it really necessary?

**Review by niklasmaki — CHANGES_REQUESTED** [review]
Aside from the one question I had, I think the code looks fine. However, I'm unable to get the clustering to run anymore even with a small dataset and 1 day time range. My browser just hangs indefinitely. Maybe someone else could test it, it might just be my computer.

File: components/cluster.py [hunk 10]
```diff
@@ -143,9 +141,11 @@ def update_output(start_date, end_date, session_id):
...
-    edges, df = compute_edges(data, df)
-    if edges == []:
+    edges = []
+    if not df.empty:
+        edges, df = compute_edges(data)
 
+    if edges == []:
         figures = [go.Figure()]
     else:
         figures = get_figures(edges, df)
```
Comment (niklasmaki) [inline components/cluster.py]: This logic is a bit weird in my opinion. Wouldn't something like this do the same?
```suggestion
    if df.empty:
        figures = [go.Figure()]
    else:
        edges, df = compute_edges(data)
        figures = get_figures(edges, df)
```

**Review by niklasmaki — APPROVED** [review]
I think it's ready to merge now :D

**Pull request merged**